imapclient/imapclient.py

```diff
@@ -1858,6 +1858,10 @@ class IMAPlibLoggerAdapter(LoggerAdapter):
     """Adapter preventing IMAP secrets from going to the logging facility."""
 
     def process(self, msg, kwargs):
+        # Fix a case in Python2 where UTF-8 encoded bytes reach the logger.
```

Comment:
I am still confused about this. How does it fail?

This code executes on Python 2 and prints True, False:
```
from __future__ import unicode_literals
  
s = b'hello'
print(isinstance(s, bytes))
print(s.startswith('>'))
```

Why would I need to decode?

Reply:
This line `print(s.startswith('>'))` is implicitly decoding the `bytes` using `ascii` in order to compare to the unicode string literal. If your input bytes contain any non-`ascii` characters this will raise a `UnicodeDecodeError`. Python 3 will always raise a `TypeError`, regardless of the content of the `bytes` object, which is much better.

I found at least one live case where Gmail flags are logged through this function in `utf-8` encoding. This change will handle those cases, and as `utf-8` is a superset of `ascii` it will not introduce any additional compatibility issues.

Reply:
Ok, true. What if we just did:
```
if isinstance(msg, str):
    for command in ...

return super(...)...
```

Would that still cause errors in `super()`?

Reply:
Not that I have seen, but from what I can see `imaplib` will always log `bytes` in Python 2 and `str` in Python 3. This would have the effect of negating the intended filter in Python 2.

Reply:
> `imaplib` will always log `bytes` in Python 2 and `str` in Python 3.

What about `logger.debug(out)` in `_send_literal`? https://github.com/closeio/imapclient/blob/cb2f99d7eb6d4096010d49edf659c68d7cfbbbcb/imapclient/imapclient.py#L1545-L1546

This would log `bytes` in Python 3 as well.

Reply:
Good point, I hadn't looked at the logger within imapclient itself. the call to `_send_literal` is conditional on the text being `ascii` compatible but it does nothing to avoid the type mismatch in the log adapter.

So it looks like I should remove the version check.